storage/postgres/storage_unsafe.sql

```diff
@@ -0,0 +1,164 @@
+-- LeafData table is keyed by the raw data hash.
+CREATE TABLE IF NOT EXISTS sequenced_leaf_data(
+  tree_id                   BIGINT NOT NULL,
+  sequence_number           BIGINT NOT NULL,
+  -- This is a personality specific has of some subset of the leaf data.
+  -- It's only purpose is to allow Trillian to identify duplicate entries in
+  -- the context of the personality.
+  leaf_identity_hash        BYTEA NOT NULL,
+  -- This is a MerkleLeafHash as defined by the treehasher that the log uses. For example for
+  -- CT this hash will include the leaf prefix byte as well as the leaf data.
+  merkle_leaf_hash          BYTEA NOT NULL,
+  integrate_timestamp_nanos BIGINT NOT NULL,
+  PRIMARY KEY(sequence_number)
+);
+
+CREATE INDEX SequencedLeafMerkleIdx ON sequenced_leaf_data(tree_id, merkle_leaf_hash);
+
+CREATE TABLE IF NOT EXISTS unsequenced(
+  tree_id               BIGINT NOT NULL,
+  -- The bucket field is to allow the use of time based ring bucketed schemes if desired. If
+  -- unused this should be set to zero for all entries.
+  bucket                INTEGER NOT NULL,
+  -- This is a personality specific hash of some subset of the leaf data.
+  -- It's only purpose is to allow Trillian to identify duplicate entries in
+  -- the context of the personality.
+  leaf_identity_hash    BYTEA NOT NULL,
+  -- This is a MerkleLeafHash as defined by the treehasher that the log uses. For example for
+  -- CT this hash will include the leaf prefix byte as well as the leaf data.
+  merkle_leaf_hash      BYTEA NOT NULL,
+  queue_timestamp_nanos BIGINT NOT NULL,
+  -- This is a SHA256 hash of the TreeID, LeafIdentityHash and QueueTimestampNanos. It is used
+  -- for batched deletes from the table when trillian_log_server and trillian_log_signer are
+  -- built with the batched_queue tag.
+  queue_id              BYTEA DEFAULT NULL UNIQUE,
+  PRIMARY KEY (queue_timestamp_nanos, leaf_identity_hash)
+);
+
+CREATE OR REPLACE FUNCTION public.ignore_duplicates(tree_id bigint, leaf_identity_hash bytea, leaf_value bytea, extra_data bytea, queue_timestamp_nanos bigint)
+ RETURNS boolean
+ LANGUAGE plpgsql
+AS $function$
+    begin
+        INSERT INTO leaf_data(tree_id,leaf_identity_hash,leaf_value,extra_data,queue_timestamp_nanos) VALUES (tree_id,leaf_identity_hash,leaf_value,extra_data,queue_timestamp_nanos);
+        return true;
+    exception
+        when unique_violation then
+                return false;
+        when others then
+                raise notice '% %', SQLERRM, SQLSTATE;
+    end;
+$function$
+
+CREATE OR REPLACE FUNCTION public.ignore_duplicates(tree_id bigint, leaf_identity_hash bytea, merkle_leaf_hash bytea, queue_timestamp_nanos bigint)
+ RETURNS boolean
+ LANGUAGE plpgsql
+AS $function$
+    begin
+        INSERT INTO unsequenced(tree_id,bucket,leaf_identity_hash,merkle_leaf_hash,queue_timestamp_nanos) VALUES(tree_id,0,leaf_identity_hash,merkle_leaf_hash,queue_timestamp_nanos);
+        return true;
+    exception when others then
```

Comment:
should this handle `unique_violation` separately from `others` (just noting the difference between this func and the one above) ?

Reply:
can do.